File: packages/singing/src/types/request.ts
import type { PitchBackendId, SeparatorBackendId } from '../constants/model-backends'

/**
 * Request body for POST /jobs/cover
 */
export interface CreateCoverRequest {
  inputUri: string
  mode: 'rvc' | 'seedvc'
  /** Original filename from the user's upload (preserved for manifest) */
  originalFileName?: string
  separator: {
    backend: SeparatorBackendId
    model?: string
    /** When true (default), isolate lead vocals from backing vocals via Karaoke model */
    isolateLeadVocal?: boolean
  }
  pitch: {
    backend: PitchBackendId
  }
  converter: CoverConverterParams
  mix?: MixParams
  /** When true (default), auto-calibrate parameters and run post-inference validation gate */
  autoCalibrate?: boolean
}

/** RVC converter parameters — field names aligned with RVC CLI/API */
export interface RvcConverterParams {
  backend: 'rvc'
  voiceId: string
  f0UpKey?: number
  indexRate?: number
  filterRadius?: number
  protect?: number
  rmsMixRate?: number
}

/** Seed-VC converter parameters — field names aligned with Seed-VC CLI */
export interface SeedVcConverterParams {
  backend: 'seedvc'
  referenceUri: string
  checkpoint?: string
  diffusionSteps?: number
  f0Condition?: boolean
  autoF0Adjust?: boolean
  semiToneShift?: number
}

export type CoverConverterParams = RvcConverterParams | SeedVcConverterParams

/** FFmpeg post-processing / remix parameters */
export interface MixParams {
  vocalGainDb?: number
  instGainDb?: number
  ducking?: boolean
  targetLufs?: number
  truePeakDb?: number
  /** Per-role gain overrides for the dynamic multi-track mixer (e.g. { backing: 0.6 }) */
  trackGains?: Record<string, number>
}

/**
 * Request body for POST /jobs/train-rvc-singer
 */
export interface CreateTrainRequest {
  voiceId: string
  datasetUri: string
  epochs?: number
  batchSize?: number
}

File: packages/singing/python/src/airi_singing_worker/dto/request.py
"""Request DTOs for the singing worker."""

from dataclasses import dataclass, field
from typing import Optional


@dataclass
class CoverRequest:
    """Request to run a cover pipeline."""

    input_path: str
    output_dir: str
    separator_backend: str = "melband"
    separator_model: str = "melband-roformer-kim-vocals"
    pitch_backend: str = "rmvpe"
    converter_backend: str = "rvc"
    voice_id: Optional[str] = None
    reference_path: Optional[str] = None
    f0_up_key: int = 0
    index_rate: float = 0.75
    protect: float = 0.33
    rms_mix_rate: float = 0.25
    # Seed-VC params
    diffusion_steps: int = 40
    f0_condition: bool = True
    auto_f0_adjust: bool = False
    semi_tone_shift: int = 0
    # Mix params
    vocal_gain_db: float = 0.0
    inst_gain_db: float = -1.5
    ducking: bool = True
    target_lufs: float = -14.0
    true_peak_db: float = -1.5
